[Inductor] Add a config flag to allow match concat_linear pattern first

zhuhaozhe opened the pull request:
The fusion pattern for `concat-linear` and `linear-pointwise` are in conflict.
For example
```
   linear(x)  -> sigmoid
   linear(x)  -> relu
```
Can either be fused as
```
                          ->relu
   concat_linear(x)->split
                          ->sigmoid
```
or
```
   linear_relu(x)
   linear_sigmoid(x)
```
So we add a flag here to make the user choose `concat linear` or 'linear point-wise` depending on the model topology.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115915



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

zhuhaozhe committed:
change pattern order to enable concat linear first

[ghstack-poisoned]

zhuhaozhe committed:
Update on "change pattern order to enable concat linear first"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "change pattern order to enable concat linear first"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "change pattern order to enable concat linear first"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "[Inductor] Add a config flag to allow match concat_linear pattern first"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[WIP][Inductor] Add a config flag to allow match concat_linear pattern first"

The fusion pattern for `concat-linear` and `linear-pointwise` are in conflict.
For example
```
   linear(x)  -> sigmoid
   linear(x)  -> relu
```
Can either be fused as
```
                          ->relu
   concat_linear(x)->split
                          ->sigmoid
```
or
```
   linear_relu(x)
   linear_sigmoid(x)
```
So we add a flag here to make the user choose `concat linear` or 'linear point-wise` depending on the model topology.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[Inductor] Add a config flag to allow match concat_linear pattern first"

The fusion pattern for `concat-linear` and `linear-pointwise` are in conflict.
For example
```
   linear(x)  -> sigmoid
   linear(x)  -> relu
```
Can either be fused as
```
                          ->relu
   concat_linear(x)->split
                          ->sigmoid
```
or
```
   linear_relu(x)
   linear_sigmoid(x)
```
So we add a flag here to make the user choose `concat linear` or 'linear point-wise` depending on the model topology.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 reviewed (changes requested):
Ideally, we should avoid leaving such complexity from users. Do we have performance data to support which one is better in what cases? I guess the batch size might impact the tradeoffs?

zhuhaozhe committed:
Update on "[Inductor] Add a config flag to allow match concat_linear pattern first"

The fusion pattern for `concat-linear` and `linear-pointwise` are in conflict.
For example
```
   linear(x)  -> sigmoid
   linear(x)  -> relu
```
Can either be fused as
```
                          ->relu
   concat_linear(x)->split
                          ->sigmoid
```
or
```
   linear_relu(x)
   linear_sigmoid(x)
```
So we add a flag here to make the user choose `concat linear` or 'linear point-wise` depending on the model topology.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "[Inductor] Add a config flag to allow match concat_linear pattern first"

The fusion pattern for `concat-linear` and `linear-pointwise` are in conflict.
For example
```
   linear(x)  -> sigmoid
   linear(x)  -> relu
```
Can either be fused as
```
                          ->relu
   concat_linear(x)->split
                          ->sigmoid
```
or
```
   linear_relu(x)
   linear_sigmoid(x)
```
So we add a flag here to make the user choose `concat linear` or 'linear point-wise` depending on the model topology.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "[Inductor] Add a config flag to allow match concat_linear pattern first"

The fusion pattern for `concat-linear` and `linear-pointwise` are in conflict.
For example
```
   linear(x)  -> sigmoid
   linear(x)  -> relu
```
Can either be fused as
```
                          ->relu
   concat_linear(x)->split
                          ->sigmoid
```
or
```
   linear_relu(x)
   linear_sigmoid(x)
```
So we add a flag here to make the user choose `concat linear` or 'linear point-wise` depending on the model topology.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
